### ISSUE TITLE
cache unfree packages on a tor-hidden binary cache server

### ISSUE THREAD

**Issue by milahu** [milahu/nixpkgs#104]
fix these issues

- [Petition to build and cache unfree packages on cache.nixos.org](https://discourse.nixos.org/t/petition-to-build-and-cache-unfree-packages-on-cache-nixos-org/17440)
- https://github.com/NixOS/nixpkgs/issues/83884

"the nix community" is too stupid
to run a [tor-hidden](https://www.torproject.org/) binary cache server...

so now all users of cuda (etc)
have to waste many hours of CPU time
(and many watthours of electrical energy)
just to build some stupid packages
just because some idiots are too stupid to run a tor-hidden website...

similar: [we need more tor-hidden gitea instances](https://www.reddit.com/r/onions/comments/1m7k3wn/we_need_more_torhidden_gitea_instances/)

packages

- torchWithCuda
  - [torch-bin](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/torch/bin/default.nix)
- jaxlibWithCuda
- tensorflowWithCuda
- numbaWithCuda
- theanoWithCuda
- adaptivecppWithCuda
- colmapWithCuda
- opensplatWithCuda
- xgboostWithCuda
- librealsenseWithCuda
- pyrealsense2WithCuda
- katagoWithCuda
- faissWithCuda
- ...

## just do it

> tor-hidden binary cache server

but probably that is not even necessary
and they could just cache unfree packages on a clearnet server like nixos.org
but they are too weak to "just do it"...
too much risk aversion, too much paranoia

the worst that can happen is that the copyright holders send mean letters
(cease and desist requests)
which produce costs of around 1 USD for a post stamp
to send back a signed cease and desist agreement

the situation is similar to filesharing of copyrighted movies etc
where it makes a difference in which jurisdiction the content is hosted
example:
in the netherlands, hosting "warez" is legal for non-commercial purposes
but in germany the copyright trolls can drag you to court and take your money

## workaround

<blockquote>

[arijoon](https://discourse.nixos.org/u/arijoon)

[Mar 2024](https://discourse.nixos.org/t/petition-to-build-and-cache-unfree-packages-on-cache-nixos-org/17440/63)

I understand this is around 2 years old but it seems like the issue is still ongoing especially for tensorflowWithCuda. I even tried using https://cuda-maintainers.cachix.org/ but it still forced a full recompile of everything when setting cudaSupport = true.

I went around it a different way by changing my tensorFlow derivation to use the wheel

```nix
{
  mtensorflowWithCuda = buildPythonPackage rec {
    pname = "tensorflow";
    version = "2.14.1";
    format = "wheel";

    src = fetchurl {
      name = "${pname}-${version}-py3-none-any.whl";
      url = "https://files.pythonhosted.org/packages/99/77/4f31cd29cab69ebc344a529df48b91a14543a83b6fb90efbf82db29a34be/tensorflow-2.14.1-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl";
      sha256 = "sha256-mpVcQhZO/011FzLBJ0ykvwWdtgyeI2IJjOHu1xd8P+k=";
    };

    nativeBuildInputs = [ pyPkgs.wheel ] ++ lib.optionals stdenv.isLinux [ autoPatchelfHook ];

    # tensorflow/tools/pip_package/setup.py
    propagatedBuildInputs = with pyPkgs; [
      absl-py
      # abseil-cpp
      astunparse
      flatbuffers
      gast
      google-pasta
      grpcio
      h5py
      keras-preprocessing
      numpy
      opt-einsum
      packaging
      # protobuf-python
      six
      tensorflow-estimator-bin
      termcolor
      typing-extensions
      wrapt
      scipy
      dm-tree
      # No longer in 310 packages, had to be copied
      # from upstream's 311 packages
      (pyPkgs.callPackage (import ./mldtypes.nix) { })
      (pyPkgs.callPackage (import ./keras.nix) { })
    ] ++ lib.optionals withTensorboard [
      tensorboard
    ];

    # During installation it can't find the deps provided above
    # but if we disable this, can assert the module works after
    pipInstallFlags = "--no-deps";

    postFixup = ''
      find $out -type f \( -name '*.so' -or -name '*.so.*' \) | while read lib; do
        # addOpenGLRunpath "$lib"
        echo [MANUAL] patching $lib

        patchelf --set-rpath "${cudatoolkit}/lib64:${cudatoolkit.lib}/lib:${cudnn_8_7}/lib:${nccl}/lib:$(patchelf --print-rpath "$lib")" "$lib"
      done
    '';

    doCheck = false;
  };
```

I had to override a couple of packages to get specific versions that work with TF 2.14, but it worked ok so far and did not require a full recompile.

I’ve shared this for two reasons

In case anyone else was still looking for a hacky solution without switching to virtual env and keep a full nix build
See if we can take advantage of wheels more often to avoid building from source, as great as it would be to have a build script from source, for certain components like this that aren’t cached by Hydra, its infeasible and would be great if we can pull the wheel along with its dependencies (it took me nearly a full day to get to this point, mainly stuck with getting cuda bindings to eventually work)

[full source code](https://github.com/arijoon/introtodeeplearning/blob/9d8f58dc530dafcf10230806c1d0fe70cf8cfc05/nix/python.nix#L22)

</blockquote>

## related

https://download.pytorch.org/whl/torch/
torch-2.11.0+cu130-cp313-cp313-manylinux_2_28_x86_64.whl
cu130 = cuda 13.0

https://discourse.nixos.org/t/installing-pytorch-into-a-virtual-python-environment/34720/2

> `LD_LIBRARY_PATH=/run/opengl-driver/lib python -c "import torch; ..."`

https://github.com/stites/pytorch-world
nix scripts for pytorch-related libraries
serve as a personal binary cache for a build-grid of pytorch. Please join [nix-community/nix-data](https://github.com/nix-community/nix-data) and petition to have community-hosted binary caches for a pytorch workflow.

https://github.com/nix-community/nix-data-science
Standard set of packages and overlays for data-scientists
from low-level packages like glibc to high-level packages like PyTorch

https://github.com/nix-community/nix-data-science/blob/master/overlays.nix

```nix
{
      pytorch = python-super.pytorch.override {
        openMPISupport = true;
        cudaSupport = true;
        buildNamedTensor = true;
        cudaArchList = [
          "5.0"
          "5.2"
          "6.0"
          "6.1"
          "7.0"
          "7.5"
          "7.5+PTX"
        ];
      };
```

https://pytorch.org/blog/pytorch-wheel-variants/

> The PyTorch packaging setup is my nemesis

https://github.com/mschoder/nix-cuda-template/blob/main/shell.nix

https://nixos.wiki/wiki/CUDA

<blockquote>

Cache: Using the [CUDA binary cache](https://cache.nixos-cuda.org/) is recommended! Add it to your configuration:

```nix
nix.settings = {
  substituters = [ "https://cache.nixos-cuda.org" ];
  trusted-public-keys = [ "cache.nixos-cuda.org:74DUi4Ye579gUqzH4ziL9IyiJBlDpMRn9MBN8oNan9M=" ];
};
```

</blockquote>

- https://github.com/NixOS/nixpkgs/issues/232426

> This issue tracks the feature-level work done to allow users to more quickly rebuild or download cached binaries.

---

reposted to https://www.reddit.com/r/onions/comments/1st9fyb/we_need_more_torhidden_binary_cache_servers/

**Comment by milahu** [comment]
https://chatgpt.com/c/69f60189-e9e0-83eb-9172-3e2e8e5ea927


why does NixOS Linux consider
the "CUDA Toolkit End User License Agreement (EULA)"
as an "unfree" license?

https://docs.nvidia.com/cuda/eula/index.html#cuda-toolkit-supplement-license-agreement

unfree license is bad, because unfree packages are not cached on the nixos binary cache (cache.nixos.org), so i have to build packages from source, which can take lots of CPU time, for example about 2 hours for the "cudaPackages.libnvshmem" package...

does the license actually prohibit the caching (redistribution) of binaries built from that source tree?

also, why is nvidia so stupid and makes life hard for its users? do they want to be hated? i guess at least Linus Torvalds hates Nvidia, because of their GPU drivers...


<blockquote>

Short answer: **yes, the CUDA EULA does restrict redistribution in ways that conflict with NixOS’s binary cache model**, and that’s why Nix marks it as *unfree*. But the reason is a bit subtler than just “no caching allowed”.

---

## 1) Why NixOS calls it “unfree”

Nixpkgs doesn’t just distinguish “open vs closed source”. It has stricter criteria:

* Must allow **free redistribution**
* Must allow **modification**
* Must not impose unusual field-of-use or platform restrictions

The CUDA EULA fails several of these.

### Key clauses

From NVIDIA’s own license:

* You get a **“non-transferable license, without the right to sublicense”** ([[NVIDIA Docs](https://docs.nvidia.com/cuda/archive/11.6.2/eula/index.html)][1])
* You **may not distribute the SDK except where explicitly allowed** ([[NVIDIA Docs](https://docs.nvidia.com/cuda/archive/11.6.2/eula/index.html)][1])
* Distribution is only permitted for **specific “distributable portions”** and only **as part of an application with additional functionality** ([[NVIDIA Docs](https://docs.nvidia.com/cuda/archive/11.6.2/eula/index.html)][1])

That’s the core problem:
👉 A general-purpose binary cache like `cache.nixos.org` is *not* “your application with additional functionality”.

So even if redistribution is *sometimes* allowed, it is **conditional and narrow**, not a blanket permission like an open-source license.

---

## 2) Does it explicitly forbid binary caching?

Not in a simple “you may not mirror binaries” sentence—but functionally, yes.

Here’s why:

* You **cannot redistribute the SDK as a standalone product** ([[NVIDIA Docs](https://docs.nvidia.com/cuda/archive/11.6.2/eula/index.html)][1])
* You can only redistribute certain parts, under strict conditions
* You cannot sublicense (important for mirrors/CDNs)

A Nix binary cache:

* Redistributes **exact builds** of packages (often including CUDA components)
* Does so **standalone**
* Does so **for arbitrary users**, not as part of a specific application

That combination falls **outside the granted rights**.

So Nixpkgs maintainers take the conservative legal interpretation:

> If redistribution isn’t clearly permitted → treat as unfree → don’t cache.

---

## 3) Why this leads to long builds

You already noticed the practical effect:

* Unfree packages are **not built on Hydra**
* Therefore not uploaded to `cache.nixos.org`
* Users must **build locally**

For CUDA-heavy packages, that’s brutal because:

* They bundle large proprietary blobs
* They often trigger heavy compilation (e.g. nvcc builds)

So yes—your 2-hour build is a direct consequence of licensing, not a technical limitation.

---

## 4) Why NVIDIA does this

Calling it “stupid” is emotionally understandable, but from their perspective it’s deliberate strategy:

### a) Control over distribution

They want:

* Users to download from NVIDIA (metrics, legal acceptance)
* Control over which versions are distributed
* Ability to revoke/change terms

### b) Ecosystem lock-in

CUDA is a **platform moat**:

* The EULA includes restrictions like:

  * no reverse engineering
  * no making it subject to open-source licenses ([[NVIDIA Docs](https://docs.nvidia.com/cuda/archive/11.6.2/eula/index.html)][1])
* Historically, even restrictions around non-NVIDIA platforms have appeared

This protects their competitive advantage.

### c) Legal risk management

They avoid:

* accidental relicensing via mirrors
* unclear redistribution chains
* compliance issues (export controls, etc.)

---

## 5) About Linus Torvalds & NVIDIA

You’re remembering correctly 🙂

Linus Torvalds publicly criticized NVIDIA ([famously saying “NVIDIA, fuck you” in 2012](https://www.youtube.com/watch?v=Q4SWxWIOVBM)), mainly about:

* poor Linux driver cooperation
* closed development model

That frustration is **philosophically aligned** with what you’re seeing:

* closed licenses
* restricted redistribution
* friction for Linux ecosystems like NixOS

---

## 6) Bottom line

* The CUDA EULA **doesn’t grant general redistribution rights**
* NixOS requires that for binary caching
* Therefore CUDA is marked **unfree**, and **not cached**
* The result is exactly what you’re experiencing: long local builds

---

## If you want a workaround

There are a few practical options (none perfect):

* Use a **community binary cache** that explicitly opts into unfree packages
* Run your own **private cache** (legally safer)
* Use **containers** (e.g. NVIDIA-provided images)
* Avoid CUDA where possible (e.g. ROCm, OpenCL—but that’s a tradeoff)

[1]: https://docs.nvidia.com/cuda/archive/11.6.2/eula/index.html "EULA :: CUDA Toolkit Documentation"


</blockquote>
